src/os_crypto/shared/keys.c

```diff
@@ -650,24 +648,34 @@ int OS_AddSocket(keystore * keys, unsigned int i, int sock) {
 
     snprintf(strsock, sizeof(strsock), "%d", sock);
     keys->keyentries[i]->sock = sock;
-    return OSHash_Set_ex(keys->keyhash_sock, strsock, keys->keyentries[i]);
+
+    w_mutex_lock(&keys->keytree_sock_mutex);
+    int r = rbtree_insert(keys->keytree_sock, strsock, keys->keyentries[i]) ? 2 : rbtree_replace(keys->keytree_sock, strsock, keys->keyentries[i]) ? 1 : 0;
+    w_mutex_unlock(&keys->keytree_sock_mutex);
+
+    return r;
 }
 
 // Delete socket number from keystore
 int OS_DeleteSocket(keystore * keys, int sock) {
     char strsock[16] = "";
     keyentry * entry;
+    int retval;
 
     snprintf(strsock, sizeof(strsock), "%d", sock);
+    w_mutex_lock(&keys->keytree_sock_mutex);
 
-    if (entry = OSHash_Delete_ex(keys->keyhash_sock, strsock), entry) {
+    if (entry = rbtree_get(keys->keytree_sock, strsock), entry) {
         if (sock == entry->sock) {
             entry->sock = -1;
+            rbtree_delete(keys->keytree_sock, strsock);
```

Comment:
The previous behavior removed this socket even if the socket number was not the same. It's okay?

Reply:
Done. The behavior was indeed changed after the refactoring. That's why that was not applied.